src/Storages/StorageFile.cpp

```diff
@@ -201,7 +201,19 @@ std::unique_ptr<ReadBuffer> selectReadBuffer(
 {
     auto read_method = context->getSettingsRef().storage_file_read_method;
 
-    if (S_ISREG(file_stat.st_mode) && read_method == LocalFSReadMethod::mmap)
+    /** But using mmap on server-side is unsafe for the following reasons:
+      * - concurrent modifications of a file will result in SIGBUS;
+      * - IO error from the device will result in SIGBUS;
+      * - recovery from this signal is not feasible even with the usage of siglongjmp,
+      *   as it might require stack unwinding from arbitrary place;
+      * - arbitrary slowdown due to page fault in arbitrary place in the code is difficult to debug.
+      *
+      * But we keep this mode for clickhouse-local as it is not so bad for a command line tool.
+      */
+
+    if (S_ISREG(file_stat.st_mode)
+        && context->getApplicationType() != Context::ApplicationType::SERVER
```

Comment:
Maybe it worth to throw in error instead of simply ignore it?

Reply:
But we want `storage_file_read_method=mmap` by default in `clickhouse-local`. So, the default value is still `mmap`, but in server we fallback to `pread`

Reply:
It will be better to have 2 defaults, 1 for server and 1 for local, but I am not sure if we can do it gracefully

Reply:
Oh, I see, make sense.

Though simply ignoring it also looks icky, maybe at least a `LOG_WARNING`, or having different defaults for `clickhouse-local` (I thought about enabling `allow_introspection_functions` for `clickhouse-local` by default).

Another option, for this particular setting is to add `auto`, that will depends on the application type.

Thoughts?

Reply:
I like the idea of separate defaults for server and local

Reply:
I also like it.

Reply:
- #51363